### PR TITLE
Feature/construct parser

### PR DESCRIPTION
**Pull request by hop-along-polly**
Feature changes for Issue #7 .

### CODE REVIEW

File: src/Parser.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace CSVParser;
+
+class Parser {
+
+    var $path;
```
Comment (hop-along-polly) [inline src/Parser.php]: Since I specified this project requires PHP7.1 in `Composer.json` you will need to specify the visibility of these member variables otherwise it will cause a syntax error. I would also recommend adding some PHPDoc comments so any developer that comes behind you know how those variables are supposed to represent. Unless you can provide sufficient reason to use pubic or protected as the visibility modifiers I am firmly requesting you specify them as private variables. Also note I renamed `headers_array` to just `headers` This was purely a style change because I don't think the _array added anything meaningful.

i.e.
```php
/*
 * @var $path The file path to the CSV file to be parsed.
 */
private $path;

/*
 * @var $headers The column headers for the CSV file that give context to the column values
 */
$headers;

/*
 * @var $delimiter The delimiter value that separates column values for this CSV File.
 */
$delimiter;
```

File: src/Parser.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace CSVParser;
+
+class Parser {
+
+    var $path;
+    var $header_array = array();
+    var $delimiter;
+
+    function __construct ($path, $header_array = array(), $delimiter = ",") {
```
Comment (hop-along-polly) [inline src/Parser.php]: Since you are specifying default values here you don't need to specify initial values above. It is also recommended you use single quotes for strings in PHP unless string interpolation is needed. I.e. this would change the method signature to 
```php
public function __construct($path, $headers = []), $delimiter = ',')
```

We can talk about string interpolation later but a simple example is
```php
$greeting = 'Hello';
echo "{$greeting} World!";

// Output
//    Hello World!
```

Comment (hop-along-polly) [reply src/Parser.php]: Also noticed I changed `$headers` to use the simpler [] declaration for an empty array. This is just stylistic and not a required change.

File: src/Parser.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+        if(file_exists($path)){ //verify path
+            $this -> $path = $path; //fill path
```
Comment (hop-along-polly) [inline src/Parser.php]: The GitHub Issue required you throw an exception if the filepath specified does not point to an existing file.

You need something like 
```php
if (!file_exists($path))
    throw new Exception("{$path} is not a valid file.");
```

File: src/Parser.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+            if($delimiter == "," || $delimiter == "|" || $delimiter == "\t") { //verify delimiter
```
Comment (hop-along-polly) [inline src/Parser.php]: you can simplify the check using the built in function  [in_array()](https://github.com/xavier643/fluent-csv-php-parser.git). You should also return some sort of error if an invalid delimiter was specified rather than allowing the Object to be constructed without a delimiter which could cause abnormal behaviour when trying to parse a CSV File.

File: src/Parser.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+                $this -> $delimiter = $delimiter; //fill delimiter
```
Comment (hop-along-polly) [inline src/Parser.php]: drop the spaces in front of and after the `->`please. It's stylistic but no one does it this way.

File: src/Parser.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+                if(count($header_array) > 0) { //make sure header array is not empty
+                    //idea on foreach to fill array came from: stackoverflow.com/questions/3045619/how-to-store-values-from-foreach-loop-into-an-array
+                    foreach($header_array as $header) {
+                        $this -> $header_array[] = $header;
+                    }
+                } else {
```
Comment (hop-along-polly) [inline src/Parser.php]: With this requests detailed above all of these `else's` will not be necessary. You'll find as I do code reviews that `else's`  are never actually needed. In rare cases I'll let an `else` slide if it makes the code easier to read.

File: src/Parser.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+                    echo "The header column is empty.";
+                }
+            } else {
+                echo "The delimter of $delimiter is not valid.";
+            }
+        } else {
+            echo "The path $path does not exist.";
+        }       
+    }
+}
```
Comment (hop-along-polly) [inline src/Parser.php]: This is originally my fault but I hate that little Icon GitHub adds on the last line. go ahead and add an empty line at the end of the file to get rid of it.

File: test/ParserTest.php [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    public function tearDown()
+    {
+        fclose($this->file);
+    }
+
+    /*
+     * @test
+     */
+    public function testConstruction()
+    {
```
Comment (hop-along-polly) [inline test/ParserTest.php]: Where are the unit tests for all of the edge cases of constructing a Parser object?

**Review by hop-along-polly — COMMENTED** [review]
Over all the implementation is on the right track. Since this is a skill building exercise I brought up a lot of stylistic and cultural things just to sharpen your skills as much as possible. 

The 2 biggest issues I see is the requirements of the GitHub Issue have not been completely fulfilled and there are no unit tests covering this code. If we need to go over PHPUnit and writing unit tests we should do that before this PR is merged.

File: test/ParserTest.php (hunk 8 above)
Comment (xavier643) [reply test/ParserTest.php]: I have not reached a point in which I know how to do the test.  I was first trying to make sure that an object was able to be made and using the correct terms and system.

File: src/Parser.php (hunk 6 above)
Comment (xavier643) [reply src/Parser.php]: The else statements where because I didn't know how to throw the error that you had previously stated as needing.  It was my best idea at the time.

File: src/Parser.php (hunk 4 above)
Comment (xavier643) [reply src/Parser.php]: I'll have to research this seeing as how delimiter is not an array but just a single variable.  I can understand the in_array() working on the header however.

File: src/Parser.php (hunk 3 above)
Comment (xavier643) [reply src/Parser.php]: I didn't know how to do this.  It's why there is the else statement later.

File: src/Parser.php (hunk 2 above)
Comment (xavier643) [reply src/Parser.php]: Another thing I did not realize was an option but I'll use it for simplicity.

File: test/ParserTest.php (hunk 8 above)
Comment (hop-along-polly) [reply test/ParserTest.php]: from command line you'll use `./vendor/bin/phpunit test`

File: src/Parser.php (hunk 1 above)
Comment (xavier643) [reply src/Parser.php]: I do have many comments, just not written in the way you have.  I used the // in order comment lines.  I do suspect they could use more information on a few of them through.

With no declaration in front of headers and delimiter does it auto assume private?

Comment (hop-along-polly) [reply src/Parser.php]: You do have plenty of comments. What I have suggested is just a standardized format that IDE's can integrate into intellisense. Other PHP developers will also be familiar with it.

File: src/Parser.php [hunk 9]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace CSVParser;
+
+class Parser {
+
+    /*
+     * @var $path the file path to the csv file to be parsed.
```
Comment (hop-along-polly) [inline src/Parser.php]: Closer to the expected format and I'm not exactly sure how intellisense would handle all the `@var`'s in a single comment this way but generally each `@var` is in it's own comment block just about the variable it is describing. That will also reduce cognitive friction for the developers reading your code.

File: src/Parser.php [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+class Parser {
+
+    /*
+     * @var $path the file path to the csv file to be parsed.
+     * @var $headers The column headers for the CSV file that give context to the column values.
+     * @var $delimiter the delimiter value that separates column values for this CSV file.
+     */
+
+    private $path;
+    private $headers = [];
```
Comment (hop-along-polly) [inline src/Parser.php]: The default value is not needed here.

File: src/Parser.php [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+    function __construct ($path, $headers = [], $delimiter = ',') {
+
+        /*
+         * Check that the path does exist or throws an exception
+         * Checks that the delimiter is a , | or /t
+         * Fills headers, if it's empty grabs first line from CSV instead
+         */
+
+        if(!file_exists($path)) throw new Exception("{$path} is not a valid file.");
```
Comment (hop-along-polly) [inline src/Parser.php]: Again totally stylistic
```php
$this->path = $path;
if (!file_exists($path))
    throw new Exception("{$path} is not a valid file.")
```

I might also change the message to say "{$path} does not exist."

File: src/Parser.php [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        $this->path = $path; //fill path
+
+        if($delimiter == ',' || $delimiter == '|' || $delimiter == '\t') { //verify delimiter
```
Comment (hop-along-polly) [inline src/Parser.php]:
```php
if (!in_array($delimiter, [',', '|', '\t']))
     throw new Exception("A delimiter of {$delimiter} is not supported.");
```

**Review by hop-along-polly — COMMENTED** [review]
I'll add a PR tomorrow that addresses how to run unit tests and from there you can work on filling in the tests and finalizing the constructor logic.

File: src/Parser.php [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        /*
+         * Check that the path does exist or throws an exception
+         * Checks that the delimiter is a , | or /t
+         * Fills headers, if it's empty grabs first line from CSV instead
+         */
+
+        if(!file_exists($path)) 
+            throw new \Exception("{$path} is not a valid file.");
+        $this->path = $path; //fill path
```
Comment (hop-along-polly) [inline src/Parser.php]: You should use fopen here so the Parser object keeps the file open while the object is active. This will be needed later when we start parsing rows of data since each time the file is opened `fgetcsv()` will have to start the from the beginning of the file which would be inefficient.

File: src/Parser.php [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+        if(!in_array($delimiter, [',', '|', '\t']))  //verify delimiter
+            throw new Exception("A delimiter of {$delimiter} is not supported.");    
+        $this->delimiter = $delimiter; //fill delimiter
+
+        $this->headers = $headers;
+        if(empty($headers)) 
+        {
+            /**
+             * @var path_open will open up path for reading
```
Comment (hop-along-polly) [inline src/Parser.php]: The @var comments are only needed on class member variables.

File: src/Parser.php [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+             */
+            $path_open = fopen($path, 'r');
+            /**
+             * @var data is an array of data from the csv file
+             */
+            $data = fgetcsv($path_open, 1000, $this->delimiter);
```
Comment (hop-along-polly) [inline src/Parser.php]: Since we cannot guarantee the max length will be 1000 characters I would set this to 0 which per the documentation means the max line length is not limited.

File: test/ParserTest.php [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+        /*
+         * create a temporary file and place it into $file
+         */
+        $this->file = tempnam(sys_get_temp_dir(), 'csv_parser_test'); 
+    }
+
+    public function tearDown()
+    {
+        /*
+         * removes the temporary file
```
Comment (hop-along-polly) [inline test/ParserTest.php]: This can be a single line comment using the //

**Review by hop-along-polly — COMMENTED** [review]
You need to add some more unittests.

File: src/Parser.php [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+            $path_open = fopen($path, 'r');
+            /**
+             * @var data is an array of data from the csv file
+             */
+            $data = fgetcsv($path_open, 0, $this->delimiter);
+            $this->headers = $data; //this line created a new error
+            //not using str_getcsv() because it makes array of arrays from whole CSV
+
+
+            //$this->headers = first row of CSV
```
Comment (hop-along-polly) [inline src/Parser.php]: These comments should go away before this is merged.

File: test/ParserTest.php [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+    /**
+     * @test
+     */
+    public function testConstructorWithNoHeader()
+    {
+        /**
+         * fill @file with default information to test when no header is sent in
+         */
+        $file_open = fopen($this->file, 'r+');
+        fwrite($file_open, "name,");
```
Comment (hop-along-polly) [inline test/ParserTest.php]: I'm going to play with it later but a single fwrite should have done it.

File: test/ParserTest.php [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+        fwrite($file_open, "social,");
+        fwrite($file_open, "123");
+        fclose($file_open);
+        $test_array = array('name', 'social', 123);
```
Comment (hop-along-polly) [inline test/ParserTest.php]: By convention this variable should be called `$expected`

File: src/Parser.php [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+            //line 31 will need to set path to an open resource in order to read this->path = fopen($path);
+            //will make path = to the resource.  File pointer.  function of fgetcsv in order to read in the rows.
+            //close the path when done
+        }
+    }
+
+    public function headers()
```
Comment (hop-along-polly) [inline src/Parser.php]: I know I was the one that originally called this `header()` but for clarity lets change it to `getHeaders()`

**Review by hop-along-polly — COMMENTED** [review]
A couple minor changes and then this is ready to be merged.

File: src/Parser.php [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+            $this->headers = $headers;
+        }
+    }
+
+    public function getHeaders()
+    {
+        return $this->headers;
+    }
+
+    public function setHeaders($path, $headers = [])
```
Comment (hop-along-polly) [inline src/Parser.php]: Imagine someone using the library like this
```php
$parser = new CSVParser($file);
$parser->setHeaders($file, ['col_1', 'col_2', 'col_3'])
echo $parser->getHeaders() . PHP_EOL;
```
Given the following file what are the proper header values? Which values will actually get printed out?
```csv
ID, Name, Title, Years
1, Derek, Senior, 7
2, Arthur, Associate, 2 
```

Also what happened to the header values I passed to `setHeaders()`?
The setHeaders funtion may work standalone but used along with the rest of the library will cause weird issues. It's better to just use the constructor as I defined rather than giving another way to set headers. Also what is the use case for wanting to set header values halfway through the parsing a CSV file? the set headers function would allow for me to read 0 to N lines and then set headers to something totally random.

Comment (hop-along-polly) [reply src/Parser.php]: I suppose it helps to understand that a valid CSV file only has one set of Header Values.

File: src/Parser.php (hunk 13 above)
Comment (xavier643) [reply src/Parser.php]: This has been edited, not sure when the fclose should be used if it's even needed.

File: test/ParserTest.php (hunk 19 above)
Comment (xavier643) [reply test/ParserTest.php]: changed and about to be updated

File: src/Parser.php (hunk 21 above)
Comment (xavier643) [reply src/Parser.php]: It has been removed.  It was put in because prior instruction was that setters and getters are always best practice but I can understand it might create issues.

**Review by hop-along-polly — COMMENTED** [review]
Looks good merge it.

**Review by xavier643 — APPROVED** [review]
